### PR TITLE
[apps] add preimage app

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**

### CODE REVIEW

File: packages/apps/contracts/PreimageApp.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
+pragma solidity 0.5.8;
+pragma experimental "ABIEncoderV2";
+
+import "@counterfactual/contracts/contracts/interfaces/CounterfactualApp.sol";
+
+
+contract PreimageApp {
```
Comment (snario) [inline packages/apps/contracts/PreimageApp.sol]: `PreImageApp` (capitalize the letter "I")

File: packages/apps/contracts/PreimageApp.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  struct AppState {
+    bytes32 image;
+    bool preimageRevealed;
+  }
+
+  struct Action {
+    bytes preimage;
+  }
+
+  function isStateTerminal(bytes memory encodedState)
```
Comment (snario) [inline packages/apps/contracts/PreimageApp.sol]: `calldata`

File: packages/apps/contracts/PreimageApp.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  function resolve(bytes memory encodedState)
+    public
+    pure
+    returns (bytes memory)
+  {
+    AppState memory state = abi.decode(encodedState, (AppState));
+
+    if (state.preimageRevealed) {
+      return abi.encode(1);
```
Comment (snario) [inline packages/apps/contracts/PreimageApp.sol]: enums

File: packages/apps/contracts/PreimageApp.sol (hunk 1 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/apps/contracts/PreimageApp.sol]: I don't agree - the spaced out version should be "preimage app" not "pre image app"

**Comment by snario** [comment]
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll status update on this PR?

**Pull request closed without merging**